module/os/linux/zfs/zpl_super.c

```diff
@@ -102,6 +105,40 @@ zpl_evict_inode(struct inode *ip)
 	spl_fstrans_unmark(cookie);
 }
 
+static void
+zpl_umount_begin(struct super_block *sb)
+{
+	zfsvfs_t *zfsvfs = sb->s_fs_info;
+
+	if (zfsvfs) {
+		/*
+		 * Flush out all POSIX I/Os.  Notify all waiters that they
+		 * must end, then wait for all users to drop their holds on
+		 * z_teardown_*_lock, and evict buffers.
+		 */
+		zfsvfs->z_unmounted = Z_UNMOUNT_BEGIN;
+		(void) dmu_objset_shutdown_register(zfsvfs->z_os);
+		rrm_enter(&zfsvfs->z_teardown_lock, RW_WRITER, FTAG);
+		rw_enter(&zfsvfs->z_teardown_inactive_lock, RW_WRITER);
+		rw_exit(&zfsvfs->z_teardown_inactive_lock);
+		rrm_exit(&zfsvfs->z_teardown_lock, FTAG);
+		dmu_objset_evict_dbufs(zfsvfs->z_os);
+
+		dsl_dir_cancel_waiters(zfsvfs->z_os->os_dsl_dataset->ds_dir);
+		dmu_objset_shutdown_unregister(zfsvfs->z_os);
+
+		/*
+		 * It's necessary to clear z_unmounted, unless the user wants
+		 * the filesystem to remain 'unmounted' for the purpose of
+		 * forced pool export.  Retaining this state causes the
+		 * filesystem to reject all but what's required to enable it
+		 * to remain in the mount table.
+		 */
+		if (!zfs_forced_export_unmount_enabled)
```

Comment:
I'm not sure I fully understand the consequences of this module option.  Do I understand correctly that when set it effectively allows the `zfsvfs_t` to orphaned from the underlying pool.  This allows it to remain in the mount table in an inoperable state but allows the pool itself is exported?  In which case this mount can then never be freed and is permanently "leaked"?  Can you add this module option to the `zfs-module-parameters.5` man page with a description.

Reply:
It's not leaked -- the pool still can't be exported, without unmounting, since the filesystem still retains its reference to its pool dataset.  All the flag does is makes it so active I/O users get forced off by failing their requests, so one can follow up with another forced export that will likely succeed, if that's all there is.

It's not quite the same as is possible on FreeBSD and Illumos, where VFS reassigns forced-unmounted vnodes to an altogether different "bad filesystem", so there are no references of any kind to the original filesystem.  But it can still be useful.  I've added documentation for this.

The alternative is to not offer the parameter, and disable the associated test cases on (only) Linux.